### PR TITLE
Simpler lifetime

### PR DESCRIPTION
**Pull request by tathanhdinh**

### CODE REVIEW

**Comment by th0rex** [comment]
Hello, thanks for your PR.

Unfortunately I don't think I'm going to merge this PR in this state as I have some further questions.

The reason that [this](https://github.com/zyantific/zydis-rs/blob/master/src/formatter.rs#L300) lifetime on the Formatter exists is that I initially thought that zydis would not create a copy of the string, however as I just noticed (because of this PR) it actually does, so in theory this lifetime could in fact be removed.
~~However, now that I know that, I think I would rather keep the lifetime and provide two API's, one that for strings uses the `ZydisFormatterSetProperty` and thus copies its argument and would not constrain the lifetime of the `Formatter` itself further. Another function would then exist to set the internal string views in zydis directly, so that no copy has to be made and then we would still need to constrain the lifetime of the `Formatter` itself.~~
~~Maybe this is overkill and a copy of at most 11 bytes does not matter, I don't really know. I think I will just add a non-copying version of that API and then that lifetime would still be required. I think that would satisfy your needs in that the copying version would not constrain the lifetime of the `Formatter` further and you would not need to keep the strings alive longer than the `Formatter`, and if someone wants to share the strings they would still have a way to do so.~~
After further considerations having a non-copying version for 11 bytes is not worth it, I would also merge this part of the PR (i.e. everything in formatter.rs except the stuff related to `FormatterToken`).

I don't know why you would want to add another lifetime to the `FormatterTokenIterator` [here](https://github.com/zyantific/zydis-rs/pull/20/files#diff-49a4b4d69f811d5619c481802227b86eR88), the only correct usage of the `FormatterToken` inside it is when both lifetimes are exactly the same, both need to have the exact lifetime that the `buffer` passed in `tokenize_operand` or in `tokenize_instruction` has. Can you explain why you added another lifetime here?

[This change](https://github.com/zyantific/zydis-rs/pull/20/files#diff-49a4b4d69f811d5619c481802227b86eR116) is fine.

I don't think I can selectively merge only some changes of the PR.
Depending on how fast you want this to be merged you can either remove the changes related to [this](https://github.com/zyantific/zydis-rs/pull/20/files#diff-49a4b4d69f811d5619c481802227b86eR88) in another commit and then I'll merge the other two things and we continue the discussion in a new PR, or we'll discuss these changes here and delay merging the other two parts until that discussion is finished.

**Comment by tathanhdinh** [comment]
Hello @th0rex and many thanks for the response,

> the only correct usage of the FormatterToken inside it is when both lifetimes are exactly the same, both need to have the exact lifetime that the buffer passed in tokenize_operand or in tokenize_instruction has

Sorry, I've forgotten lifetime notations for the output references. Hope the new update makes you happy. 

> I don't know why you would want to add another lifetime to the FormatterTokenIterator

That's because I think that any reference to `FormatterToken<'b>` should keep the lifetime of the buffer, otherwise the following pseudo code should not compile:

```rust
let buffer = ...;  // lifetime 'b
{
   let ft;         // lifetime 'a
   {
      ft = tokenize_operand(..., buffer, ...).unwrap();
   }
   let fi = FormatterTokenIterator { next: Some(ft) };  // next: Option<&'a FormatterToken<'a>>
   // 'b is lost: cannot use fi to reference something with lifetime 'b
}
```

**Comment by tathanhdinh** [comment]
Yooo, PR updated. Sorry for the delay :(

File: src/ffi.rs [hunk 1]
```diff
@@ -53,7 +53,7 @@ impl<'a> FormatterToken<'a> {
             let mut val = mem::uninitialized();
             check!(ZydisFormatterTokenGetValue(self, &mut ty, &mut val))?;
 
-            let val = CStr::from_ptr(val as *const i8)
+            let val = CStr::from_ptr(val as _)
```
Comment (th0rex) [inline src/ffi.rs:56]: I think `*const _` is nicer, you've also done so below. I'm just going to merge your PR and change it myself.

**Review by th0rex — COMMENTED** [review]
Thanks for the updates and the PR!

**Pull request merged**